File: elementary_data_structures/queues.py
class Element:

    __slots__ = ['data', 'next']

    def __init__(self, data):
        self.data = data
        self.next = None


class QueueEmptyException(Exception):

    pass


class Queue:

    __slots__ = ['size', 'head', 'tail']

    def __init__(self):
        self.size = 0
        self.head = self.tail = None

    def enqueue(self, z):
        self.size += 1

        if not self.head:
            self.head = z
        else:
            self.tail.next = z
        self.tail = z

    def dequeue(self):
        if self.size == 0:
            raise QueueEmptyException

        self.size -= 1

        dequeued = self.head
        result = dequeued.data
        self.head = dequeued.next
        dequeued = None

        return result


    def print_queue(self):
        print(f"size = {self.size} : ", end='')

        p = self.head
        while p:
            print(f"{p.data}", end=' -> ' if p.next else '\n')

            p = p.next


if __name__ == '__main__':
    q = Queue()

    while True:
        print("1:enqueue 2:dequeue 3:print > ", end='')
        op = int(input())

        if op == 1:
            print("input data > ", end='')
            data = int(input())

            q.enqueue(Element(data))
        elif op == 2:
            try:
                result = q.dequeue()

                print(f"result = {result}")
            except QueueEmptyException:
                print("queue is empty.")
        elif op == 3:
            q.print_queue()
        else:
            print("invalid operation")

File: elementary_data_structures/queues.go
//go:build ignore

package main

import "fmt"

type Element struct {
	data int
	next *Element
}

type Queue struct {
	size int
	head *Element
	tail *Element
}

func (q *Queue) enqueue(z *Element) {
	q.size += 1

	if q.head == nil {
		q.head = z
	} else {
		q.tail.next = z
	}
	q.tail = z
}

func (q *Queue) dequeue() int {
	if q.size == 0 {
		fmt.Println("queue is empty.")

		return -1
	}

	q.size -= 1

	e := q.head
	result := e.data
	q.head = e.next
	e = nil

	return result
}

func (q *Queue) printQueue() {
	fmt.Printf("size = %d : ", q.size)

	p := q.head
	for p != nil {
		fmt.Printf("%d", p.data)
		if p.next != nil {
			fmt.Print(" -> ")
		}

		p = p.next
	}

	fmt.Println()
}

func main() {
	q := Queue{0, nil, nil}

	for {
		fmt.Print("1:enqueue 2:dequeue 3:print > ")
		var op int
		fmt.Scanf("%d", &op)

		switch op {
		case 1:
			fmt.Print("input data > ")
			var data int
			fmt.Scanf("%d", &data)

			z := Element{data, nil}
			q.enqueue(&z)
		case 2:
			result := q.dequeue()
			if result != -1 {
				fmt.Printf("result = %d\n", result)
			}
		case 3:
			q.printQueue()
		default:
			fmt.Println("invalid operation")
		}
	}
}
